w3lib/url.py

```diff
@@ -95,10 +95,29 @@ def safe_url_string(
     decoded = to_unicode(url, encoding=encoding, errors="percentencode")
     parts = urlsplit(_ascii_tab_newline_re.sub("", decoded))
 
+    username, password, hostname, port_number = (
+        parts.username,
+        parts.password,
+        parts.hostname,
+        parts.port,
+    )
+    netloc_bytes = b""
+
     # IDNA encoding can fail for too long labels (>63 characters)
     # or missing labels (e.g. http://.example.com)
     try:
-        netloc_bytes = parts.netloc.encode("idna")
+        # When we have hostname we use it instead of netloc directly
+        if hostname:
+            if isinstance(username, str) and isinstance(password, str):
+                netloc_bytes += quote(username, _safe_chars).encode(encoding)
```

Comment:
Do we need `quote` here? I’m mostly concerned because `_safe_chars` contains some characters that may not be safe here, such as `:` for a username.

Reply:
I have the same question, i follow the SO answer, what do you think? Or maybe we could remove the `_safe_chars` param.

Reply:
@Gallaecio Do you think worth add test for empty username or this is a edgy case?